src/ethereum-json-rpc-client/src/lib.rs

```diff
@@ -1,4 +1,257 @@
+        method: String,
+        params: impl IntoIterator<Item = (Params, Id)>,
+        max_batch_size: usize,
+    ) -> anyhow::Result<Vec<R>> {
+        let mut results = Vec::new();
+
+        let value_from_json = |value| serde_json::from_value::<R>(value);
+
+        // Collect chunks before iteration, otherwise the future won't be `Send`
+        let chunks = params
+            .into_iter()
+            .chunks(max_batch_size)
+            .into_iter()
+            .map(Iterator::collect::<Vec<_>>)
+            .collect::<Vec<_>>();
+        for chunk in chunks {
+            let method_calls = chunk
+                .into_iter()
+                .map(|(params, id)| {
+                    Call::MethodCall(MethodCall {
+                        jsonrpc: Some(Version::V2),
+                        method: method.clone(),
+                        params,
+                        id,
+                    })
+                })
+                .collect::<Vec<_>>();
+            let chunk_size = method_calls.len();
+            let request = Request::Batch(method_calls);
+
+            let response = self.client.send_rpc_request(request).await?;
+
+            match response {
+                Response::Single(response) => match response {
+                    Output::Success(result) => {
+                        if chunk_size == 1 {
+                            results.push(value_from_json(result.result)?);
+                        } else {
+                            anyhow::bail!(
+                                "unexpected number of results: have: 1, expected {chunk_size}"
+                            );
+                        }
+                    }
+                    Output::Failure(err) => {
+                        anyhow::bail!("{err:?}");
+                    }
+                },
+                Response::Batch(response) => {
+                    if chunk_size == response.len() {
+                        for resp in response.into_iter() {
+                            match resp {
+                                Output::Success(resp) => {
+                                    results.push(value_from_json(resp.result)?)
+                                }
+                                Output::Failure(err) => {
+                                    anyhow::bail!("{err:?}");
+                                }
+                            }
+                        }
+                    } else {
+                        anyhow::bail!(
+                            "unexpected number of results: have: {}, expected {}",
+                            response.len(),
+                            chunk_size
+                        );
+                    }
+                }
+            }
+        }
+
+        Ok(results)
+    }
+}
+
+pub trait Client: Clone + Send + Sync {
+    fn send_rpc_request(
+        &self,
+        request: Request,
+    ) -> Pin<Box<dyn Future<Output = anyhow::Result<Response>> + Send + Sync>>;
```

Comment:
Why don't use async_trait for it?

Reply:
This is the same as what `async_trait` does under the hood with the difference that in this case I have better control over the `Send`/`Sync` behavior. With `async_trait`, in fact, I wan't able to fix a compilation error in the evm-canister due to that

Reply:
Yes, that's weird. `async_trait` supports disabling `Send` and `Sync` manually and they are on by default